# Evaluating Goodness of Fit of Linear Regression Models

So far, we've learned how to build a linear regression model and interpret the estimated coefficients. In this checkpoint, we discuss how we evaluate the performance of our models in the training phase. Recall that we can talk about two broad types of performances: one in the training set and the other in the test set. The former enables us to talk about how well our model explains the information in the target variable and the later gives us an idea about how well our model will perform when it's given previously unseen observations.

First, we start with the investigation of whether an estimated model deems better than a reduced model. By reduced model, we mean a model with no features. To compare our model with the reduced model, we use **F-test**.

Second, we delve into the measurement of how well our model explains the variance in the target variable. To this end, we talk about **R-squared** and **adjusted R-squared**.

Last, we explain how we can compare different models in terms of their explanatory power. We show how to read **Akaike** and **Bayesian** information criterias for this purpose.

## Is our model better than an "empty" model?

When evaluating our model, we first need to ask whether our model contributes anything to the explanation of the outcome variable. In other words, we need to determine whether our features are any useful in explaining the variance of the outcome. If not, we can drop our features altogether and the resulting "empty" model would do the same job.

For this purpose, we use **F-test**.

#### The F-test

The F-test can be calculated in different ways depending on the situation, but principally represents the ratio between the unexplained variance of our model and the unexplained variance of a reduced model to which our model is compared.  Here, the "reduced model" is a model with no features, meaning all variance in the outcome is unexplained.  For a linear regression model with two parameters $y=\alpha+\beta x$, the F-test is built from these pieces:

* unexplained model variance:

$$SSE_F=\sum(y_i-\hat{y}_i)^2$$

* unexplained variance in reduced model:

$$SSE_R=Var_y = \sum(y_i-\bar{y})^2$$

 * number of parameters in the model:

$$p_F = 2 (\alpha \text{ and } \beta)$$

 * number of parameters in the reduced model:

$$p_R = 1 (\alpha)$$

 * number of datapoints:

$$n$$

 * degrees of freedom of $SSE_F$:

$$df_F = n - p_F$$

 * degrees of freedom of $SSE_R$:

$$df_R = n - p_R$$

These pieces come together to give us the full equation for the F-test:

$$F=\dfrac{SSE_F-SSE_R}{df_F-df_R}÷\dfrac{SSE_F}{df_F}$$

This introduces some new terminology. **Degrees of freedom** quantify the amount of information "left over" to estimate variability after all parameters are estimated.

In regression, degrees of freedom for a function works like this:  With two datapoints, a regression line $y=\alpha + \beta x$ has 0 degrees of freedom (2 minus the number of parameters).  Those two parameters encompass all the information in the data.  Knowing $\alpha$ and $\beta$ alone, we can perfectly reproduce the original data.  No additional information is available from the data itself.

The null hypotheses of the F-test states that the model is indifferent from the reduced model which means that the features contributes nothing to the explanation of the target variable. Instead of reading the F statistics, it's easier to read the p-value of it. The lesser the p-value, the better for our model. Namely, if the p-value of the F test for our model is less than or equal to 0.1 (or even less than or equal to 0.05), we say that our model is useful and contributes something that is statistically significant in the explanation of the target.

Let's see the F statistic of our medical costs model:

In [8]:
import math
import warnings

from IPython.display import display
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sqlalchemy import create_engine

# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

# Suppress annoying harmless error.
warnings.filterwarnings(action="ignore")

In [9]:
engine = create_engine('postgresql://username:pass@localhost:5432/MedicalCosts')

insurance_df = pd.read_sql_query('select * from medicalcosts',con=engine)
insurance_df.head(10)

age     sex    bmi  children smoker     region   charges
0   19  female 27.900         0    yes  southwest 16884.924
1   18    male 33.770         1     no  southeast  1725.552
2   28    male 33.000         3     no  southeast  4449.462
3   33    male 22.705         0     no  northwest 21984.471
4   32    male 28.880         0     no  northwest  3866.855
5   31  female 25.740         0     no  southeast  3756.622
6   46  female 33.440         1     no  southeast  8240.590
7   37  female 27.740         3     no  northwest  7281.506
8   37    male 29.830         2     no  northeast  6406.411
9   60  female 25.840         0     no  northwest 28923.137

In [10]:
insurance_df["is_male"] = pd.get_dummies(insurance_df.sex, drop_first=True)
insurance_df["is_smoker"] = pd.get_dummies(insurance_df.smoker, drop_first=True)

insurance_df.region[np.where(np.isin(insurance_df.region, "southwest"))[0]] = 0
insurance_df.region[np.where(np.isin(insurance_df.region, "northwest"))[0]] = 1
insurance_df.region[np.where(np.isin(insurance_df.region, "southeast"))[0]] = 2
insurance_df.region[np.where(np.isin(insurance_df.region, "northeast"))[0]] = 3

In [11]:
# Y is the target variable
Y = insurance_df['charges']

# X is the feature set
X = insurance_df[['is_male','is_smoker', 'age', 'bmi']]

# We add constant to the model as it's a best practice
# to do so everytime!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results.
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.747
Model:                            OLS   Adj. R-squared:                  0.747
Method:                 Least Squares   F-statistic:                     986.5
Date:                Thu, 18 Oct 2018   Prob (F-statistic):               0.00
Time:                        17:22:21   Log-Likelihood:                -13557.
No. Observations:                1338   AIC:                         2.712e+04
Df Residuals:                    1333   BIC:                         2.715e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -1.163e+04    947.267    -12.281      0.0

Model's F statistic is 986.5 and the associated p-value is very close to zero. This means that, our features add some information to the reduced model and our model is useful for explaning the charges.

However, F-test doesn't provide us a measure to quantify how much information our model contributes. To this purpose, we discuss the R-squared next.

## Quantifying the performance of a model in the training set

R-squared is probably the most common measure of goodness of fit in a linear regression model. It is a proportion (between 0 and 1) that expresses how much variance in the outcome variable is explained by the explanatory variables in the model. Generally speaking, higher $R^2$ values are better to a point-- a low $R^2$ indicates that our model isn't explaining much information about the outcome, which means it will not give very good predictions. However, a very high $R^2$ is a warning sign for overfitting.  No dataset is a perfect representation of reality, so a model that perfectly fits our data ($R^2$ of 1 or close to 1) is likely to be biased by quirks in the data, and will perform less well on the test-set.

In the regression summary table above, we see that the R-squared value of our medical costs model is 0,747. It means that our model explains 74,7% of the variance in the charges. So, 25,3% is still unexplained. Hence, we can conclude that there are still room for improvement. Let's fit the model in the previous checkpoint again where we included the interaction of body mass index (bmi) and is_smoking dummy:

In [12]:
# Y is the target variable
Y = insurance_df['charges']

# This is the interaction between bmi and smoking
insurance_df["bmi_is_smoker"] = insurance_df.bmi * insurance_df.is_smoker

# X is the feature set
X = insurance_df[['is_male','is_smoker', 'age', 'bmi', "bmi_is_smoker"]]

# We add constant to the model as it's a best practice
# to do so everytime!
X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.837
Model:                            OLS   Adj. R-squared:                  0.836
Method:                 Least Squares   F-statistic:                     1365.
Date:                Thu, 18 Oct 2018   Prob (F-statistic):               0.00
Time:                        18:11:24   Log-Likelihood:                -13265.
No. Observations:                1338   AIC:                         2.654e+04
Df Residuals:                    1332   BIC:                         2.657e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const         -2071.0773    840.644     -2.464

The R-squared of this model is 0,837 which is higher than our previous model's. This improvement indicates that the interactions of bmi and is_smoker explains some previously unexplained variance of the charges. 

As we said before, high R-squared values are generally something admirable. However, in some cases, very high R-squared values indicate some potential problems with our model. Specifically:

* Very high R-squared value may be a sign of overfitting. If our model is too complex for the data, then it may overfit the training set and does a poor job in the test set. **To assess this, we need to evaluate the performance of our model in the test set and see whether our model does a significantly poor job in the test set compared to the training set**. We'll discuss how to evaluate liear regression models in the test set in the next checkpoint.

* R-squared is inherently a biased estimate of the performance in the sense that the more explanatory variables are added to the model, the higher R-squared values we get. This is so even we include irrelevant variables like noises or random data. To mitigate this problem, we usually use a metric called **adjusted R-squared** instead of R-squared. Adjusted R-squared does the same job with the R-squared but it is adjusted according to the number of features included to the model. Hence, it's always safer to look at the adjusted R-squared value instead of R-squared value.

## Comparing different models

Comparing different models and chosing the best one is one of the essential practices in Data Science. Often, we try several models and evaluate them in the test set in order to detect the top performing one. However, *inference* is also a very critical task when it comes to linear regression models. Unlike testing the predictive power, in inference, we care about the explanatory power of our models.

Throughout this checkpoint, we saw that we can measure the performance of our models in the training set using F-test or R-squared. Hence, both F-test and R-squared can be used in the comparison of different models. Unfortunately, the two metrics suffer from some drawbacks which make them inappropriate to use in some situations when comparing different models.

Here we briefly overview how we can use F-test and R-squared in model comparison. Then, we introduce information criterions that we can also use to compare different models.

#### Using F-test for model comparison

We can use F-test to compare two models if one of them is nested within the other. That is, if the feature set in a model is a subset of the feature set of the other, then we can use F-test. In this case, we say that the model with higher F statistics is superior to the other one.

However, if models are not nested, then using F-test might be misleading. It's quite sensitive to the normality of the error terms. If errors are not normally distributed, we should try other methods.

#### Using R-squared for model comparision

Another metric we can use for model comparison is the R-squared. We already saw that R-squared is biased as it tends to increase with the number of explanatory variables. So, instead of R-squared we can use adjusted R-squared. The higher adjusted R-squared, the better model explains the target variable. 

#### Using information criterions

Using information criterions is also a common way of comparing different models and selecting the best one. Here, we talk about two information criterias known as **Akaike Information Criteria (AIC)** and **Bayesian Information Criterian (BIC)**. Both of these information criterias take into consideration the sum of the squared errors (SSE), the sample size and the number of parameters.

The formula for AIC is:

$$nln(SSE)−nln(n)+2p$$ 


and the formula for BIC is:

$$nln(SSE)−nln(n)+pln(n)$$

In both of these formulas, $n$ represents the sample size and $p$ represents the number of regression coefficients in the model (including the constant).

For both of AIC and BIC, the lower value the better. Hence, we choose the model with the lowest AIC or BIC value. Although, we can use ony one of the two criterias, AIC is usually criticized for its tendency to overfit. In contrast, BIC penalizes the number of parameters more severely than AIC and hence favors more parsimonious (models with lesser number of parameters) models.

#### Which medical costs model is better?

statmodels' `summary()` function gives us all of the above metrics. In the tables above, we see that for our first model, R-squared is 0.747, adjusted R-squared is 0.747, F statistics is 986.5, AIC is 27.120 and BIC is 27.150 whereas for our second model, R-squared is 0.837, adjusted R-squared is 0.836, F statistics is 1365, AIC is 26.540 and BIC is 26.570. According to all of the metrics, our second model seems better than the first one.

# Assignment

Using your house prices model:

* Assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.
* Do you think your model is satisfactory? If so, why?
* In order to improve the goodness of fit of your model, try different model specifications by adding and removing some variables. 
* For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?